src/libs/ease/hash.c

```diff
@@ -0,0 +1,118 @@
+ *
+ * The KeySet of an application is identified as all keys below the applications root key.
+ *
+ * @pre The parentKey must have the correct namespace. E.g. If only keys from the spec:/ should be considered for the token calculation,
+ * pass a key with KEY_NS_SPEC.
+ *
+ * @param hash_string A string. After successful execution this will contain the hash as hex-string.
+ * @param ks The KeySet for the application.
+ * @param parentKey The Key below which all the relevant keys are. Keys that are not below @p parentKey are ignored. The key's namespace is important (see preconditions)
+ * @retval false If an error occurred.
+ * @retval true If the computation was successful.
+ */
+kdb_boolean_t calculateSpecificationToken (char hash_string[65], KeySet * ks, Key * parentKey) {
+	if(parentKey == NULL) {
+		// Can't set error to parentKey when it is null.
+		return false;
+	}
+	if(hash_string == NULL) {
+		ELEKTRA_SET_INTERNAL_ERROR(parentKey, "Param hash_string was NULL");
+		return false;
+	}
+	if(ks == NULL) {
+		ELEKTRA_SET_INTERNAL_ERROR(parentKey, "Param ks was NULL");
+		return false;
+	}
+
+	// Initialize sha_256 for streaming
+	uint8_t hash[SIZE_OF_SHA_256_HASH];
+	struct Sha_256 sha_256;
+	sha_256_init(&sha_256, hash);
+
+	// Duplicate ks, then cut out parentKey and all keys below. These are the ones we take into account for token calculation.
+	KeySet * dupKs = ksDup (ks);
+	KeySet * cutKs = ksCut (dupKs, parentKey);
+
+	/**
+	 * Loop through all keys relevant for token calculation.
+ 	*/
+	Key * currentKey;
+	ksRewind(cutKs);
+	while ((currentKey = ksNext (cutKs)) != NULL) {
+		Key * cascadingKey = keyDup (currentKey, KEY_CP_NAME);
+		if(cascadingKey == NULL) {
+			ELEKTRA_SET_INTERNAL_ERROR (parentKey, "keyDup() failed!");
+			return false;
+		}
+		/**
+		 * Key namespace is ignored for token calculation (via setting it to cascading).
+		 * Reason: Different callers of this function pass the keys for token calculation using different namespaces,
+		 * but the specification they pass is actually the same.
+		 * (e.g. tools/kdb/gen.cpp passes keys in cascading namespace while src/libs/highlevel/elektra.c passes keys in spec namespace).
+		 */
+		int result = keySetNamespace (cascadingKey, KEY_NS_CASCADING);
+		if(result == -1) {
+			ELEKTRA_SET_INTERNAL_ERROR (parentKey, "keySetNamespace() failed!");
+			return false;
+		}
+		// Feed key name into sha_256_write() without NULL terminator (hence -1).
+		// This makes it easier to compare expected results with other sha256 tools.
+		sha_256_write (&sha_256, keyName(cascadingKey), keyGetNameSize(cascadingKey) - 1);
+		// Note: The value of the key itself is not relevant / part of specification. Only the key's name + its metadata!
+
+		KeySet * currentMetaKeys = keyMeta(currentKey);
+		Key * currentMetaKey;
+		ksRewind(currentMetaKeys);
+		// Feed name + values from meta keys into sha_256_write().
+		while ((currentMetaKey = ksNext (currentMetaKeys)) != NULL) {
```

Comment:
Also here please use new style of iteration.

It is good that you already use `keyMeta`, how did you find it? (keyNextMeta is the old style @lawli3t can you also mark this as deprecated?)

Reply:
Thanks for pointing that out! I wasn't aware of a new vs. old style of iteration. I picked up ksNext, ksRewind and keyMeta simply by looking at existing code within libelektra.

I'll switch to the new style.

Reply:
I switched to the new style!